File: wsproxy.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"net"
	"os"
	"syscall"
	"time"
)

// Standard HTTP 101 Switching Protocols payload
var http101Response = []byte(
	"HTTP/1.1 101 Switching Protocols\r\n" +
		"Upgrade: websocket\r\n" +
		"Connection: Upgrade\r\n\r\n",
)

// Optimize TCP socket flags directly at kernel level
func optimizeSocket(conn net.Conn) {
	tcpConn, ok := conn.(*net.TCPConn)
	if !ok {
		return
	}

	// Disable Nagle's algorithm for instant packet delivery
	_ = tcpConn.SetNoDelay(true)

	// Set low-level socket options on Linux
	rawConn, err := tcpConn.SyscallConn()
	if err != nil {
		return
	}

	_ = rawConn.Control(func(fd uintptr) {
		// Enable TCP_QUICKACK (12) to force immediate ACKs and eliminate ~40ms delays
		_ = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, 12, 1)
	})
}

func handleClient(clientConn net.Conn) {
	defer clientConn.Close()
	optimizeSocket(clientConn)

	// Read initial HTTP WebSocket upgrade GET request header
	buf := make([]byte, 4096)
	n, err := clientConn.Read(buf)
	if err != nil || n == 0 {
		return
	}

	// Reply immediately with HTTP 101 Switching Protocols header
	if _, err := clientConn.Write(http101Response); err != nil {
		return
	}

	// Dial local SSH daemon directly on port 22
	sshConn, err := net.DialTimeout("tcp", "127.0.0.1:22", 3*time.Second)
	if err != nil {
		return
	}
	defer sshConn.Close()
	optimizeSocket(sshConn)

	// Extract trailing payload bytes if client bundled SSH data in initial handshake packet
	headerEnd := bytes.Index(buf[:n], []byte("\r\n\r\n"))
	if headerEnd != -1 && headerEnd+4 < n {
		payload := buf[headerEnd+4 : n]
		if _, err := sshConn.Write(payload); err != nil {
			return
		}
	}

	// Bidirectional Kernel Zero-Copy (uses splice system call on Linux)
	done := make(chan struct{}, 2)

	go func() {
		_, _ = io.Copy(sshConn, clientConn)
		done <- struct{}{}
	}()

	go func() {
		_, _ = io.Copy(clientConn, sshConn)
		done <- struct{}{}
	}()

	// Wait until either direction terminates
	<-done
}

func main() {
	// Configure high-performance TCP listener with SO_REUSEADDR and SO_REUSEPORT
	lc := net.ListenConfig{
		Control: func(network, address string, c syscall.RawConn) error {
			return c.Control(func(fd uintptr) {
				_ = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
				_ = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, 15, 1) // 15 = SO_REUSEPORT
			})
		},
	}

	listener, err := lc.Listen(nil, "tcp", "127.0.0.1:2222")
	if err != nil {
		fmt.Fprintf(os.Stderr, "[-] Failed to bind wsproxy listener: %v\n", err)
		os.Exit(1)
	}
	defer listener.Close()

	fmt.Println("[+] High-Performance Go WS Proxy running on 127.0.0.1:2222")

	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		go handleClient(conn)
	}
}

File: wsproxy.py
import asyncio
import socket

# Constant 101 HTTP Upgrade Response
HTTP_101_RESPONSE = (
    b"HTTP/1.1 101 Switching Protocols\r\n"
    b"Upgrade: websocket\r\n"
    b"Connection: Upgrade\r\n\r\n"
)

def optimize_socket(sock):
    """Apply zero-delay kernel flags to local loopback socket."""
    if not sock:
        return
    # Disable Nagle's algorithm for instant packet dispatch
    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
    
    # Disable delayed ACKs on Linux (prevents ~40ms ACK delay overhead)
    if hasattr(socket, "TCP_QUICKACK"):
        try:
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_QUICKACK, 1)
        except OSError:
            pass


class BridgeProtocol(asyncio.Protocol):
    """Zero-overhead bidirectional proxy protocol with flow control."""
    def __init__(self, target_transport=None):
        self.transport = None
        self.peer_transport = target_transport

    def connection_made(self, transport):
        self.transport = transport
        optimize_socket(transport.get_extra_info('socket'))

    def data_received(self, data):
        if self.peer_transport:
            self.peer_transport.write(data)

    def pause_writing(self):
        if self.peer_transport:
            self.peer_transport.pause_reading()

    def resume_writing(self):
        if self.peer_transport:
            self.peer_transport.resume_reading()

    def connection_lost(self, exc):
        if self.peer_transport:
            self.peer_transport.close()
            self.peer_transport = None


class ClientBridgeProtocol(BridgeProtocol):
    """Client handler with fast 101 handshake, buffer queueing, and stream bridging."""
    def __init__(self):
        super().__init__()
        self.handshake_sent = False
        self.pending_buffer = bytearray()

    def connection_made(self, transport):
        super().connection_made(transport)
        loop = asyncio.get_running_loop()
        asyncio.create_task(self._connect_ssh(loop))

    async def _connect_ssh(self, loop):
        try:
            _, ssh_protocol = await loop.create_connection(
                lambda: BridgeProtocol(target_transport=self.transport),
                '127.0.0.1', 
                22
            )
            self.peer_transport = ssh_protocol.transport
            
            # Flush any payload received while SSH socket was establishing
            if self.pending_buffer:
                self.peer_transport.write(self.pending_buffer)
                self.pending_buffer = None
        except Exception:
            self.transport.close()

    def data_received(self, data):
        if not self.handshake_sent:
            self.handshake_sent = True
            self.transport.write(HTTP_101_RESPONSE)
            
            # Find boundary of initial HTTP request (\r\n\r\n)
            header_end = data.find(b"\r\n\r\n")
            if header_end != -1:
                # Retain payload bytes attached to the initial handshake packet
                payload = data[header_end + 4:]
                if payload:
                    if self.peer_transport:
                        self.peer_transport.write(payload)
                    else:
                        self.pending_buffer.extend(payload)
            return

        if self.peer_transport:
            self.peer_transport.write(data)
        elif self.pending_buffer is not None:
            self.pending_buffer.extend(data)


async def main():
    loop = asyncio.get_running_loop()
    
    server = await loop.create_server(
        ClientBridgeProtocol,
        '127.0.0.1',
        2222,
        backlog=65535,
        reuse_port=True
    )
    print("[+] Zero-Delay WS Proxy running on 127.0.0.1:2222")
    async with server:
        await server.serve_forever()


if __name__ == '__main__':
    asyncio.run(main())
